### PR TITLE
Add Stable Diffusion 2.1 example

### PR DESCRIPTION
**Pull request by attila-dusnoki-htec**
This example show how to use HuggingFace's [Stable Diffusion v2.1](https://huggingface.co/stabilityai/stable-diffusion-2-1) model with migraphx's python frontend.

### CODE REVIEW

File: examples/diffusion/python_stable_diffusion_21/README.md [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+```bash
+# this will require the python venv to installed (e.g. apt install python3.8-venv)
+python3 -m venv sd_venv
+. sd_venv/bin/activate
+```
+
+Install dependencies
+
+```bash
+pip install optimum[onnxruntime] transformers diffusers accelerate
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/README.md:24]: can you make a requirements.txt?

File: examples/diffusion/python_stable_diffusion_21/README.md [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+*Note: `models/sd21-onnx` will be used in the scripts.*
+
+Run text to image example with the following example prompt and seed:
+
+```bash
+python txt2img.py --prompt "a photograph of an astronaut riding a horse" --seed 13 --output astro_horse.png
+```
+*Note: The first run will compile the models and cache them to make subsequent runs faster.*
+
+The result should like this:
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/README.md:47]:
```suggestion
The result should look like this:
```

File: examples/diffusion/python_stable_diffusion_21/README.md [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+This version was tested with [rocm 5.7](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/tree/rocm-5.7.0) revision.
+
+## Jupyter notebook
+
+There is a dedicated step-by-step notebook. See [sd21.ipynb](./sd21.ipynb)
+
+## Console application
+
+To run console application, follow these steps below.
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/README.md:11]:
```suggestion
To run the console application, follow these steps below.
```

File: examples/diffusion/python_stable_diffusion_21/README.md [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+```
+
+Get models with optimum
+
+```bash
+optimum-cli export onnx --model stabilityai/stable-diffusion-2-1 models/sd21-onnx
+```
+*Note: `models/sd21-onnx` will be used in the scripts.*
+
+Run text to image example with the following example prompt and seed:
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/README.md:40]:
```suggestion
Run text-to-image script with the following example prompt and seed:
```

File: examples/diffusion/python_stable_diffusion_21/sd21.ipynb [hunk 5]
```diff
@@ -0,0 +1,532 @@
...
+    "!optimum-cli export onnx --model stabilityai/stable-diffusion-2-1 models/sd21-onnx"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Now it is time to load these models with python.\n",
+    "\n",
+    "First, we make sure that MIGraphX is in the path."
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/sd21.ipynb:81]:
```suggestion
    "First, we make sure that MIGraphX module is found in the python path."
```

File: examples/diffusion/python_stable_diffusion_21/sd21.ipynb [hunk 6]
```diff
@@ -0,0 +1,532 @@
...
+    "        print(f\"Found mxr, loading it...\")\n",
+    "        model = mgx.load(f\"{file}.mxr\", format=\"msgpack\")\n",
+    "    elif os.path.isfile(f\"{file}.onnx\"):\n",
+    "        print(f\"Parsing from onnx file...\")\n",
+    "        model = mgx.parse_onnx(f\"{file}.onnx\", map_input_dims=shapes)\n",
+    "        model.compile(mgx.get_target(\"gpu\"))\n",
+    "        print(f\"Saving {name} model to mxr file...\")\n",
+    "        mgx.save(model, f\"{file}.mxr\", format=\"msgpack\")\n",
+    "    else:\n",
+    "        print(f\"No {name} model found. Please download it and re-try.\")\n",
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/sd21.ipynb:127]:
```suggestion
    "        print(f\"No {name} model found. Please verify the path is correct and re-try, or re-download model.\")\n",
```

File: examples/diffusion/python_stable_diffusion_21/sd21.ipynb [hunk 7]
```diff
@@ -0,0 +1,532 @@
...
+    "        os.exit(1)\n",
+    "    return model"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "With that, we can load the models. This could take minutes."
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/sd21.ipynb:136]:
```suggestion
    "With that, we can load the models. This could take several minutes."
```

File: examples/diffusion/python_stable_diffusion_21/sd21.ipynb [hunk 8]
```diff
@@ -0,0 +1,532 @@
...
+    "def compute_previous(noise_pred, t, latents):\n",
+    "    # compute the previous noisy sample x_t -> x_t-1\n",
+    "    return scheduler.step(noise_pred, t, latents).prev_sample\n"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "Scale and decode the image latents with vae."
```
Comment (kahmed10) [inline examples/diffusion/python_stable_diffusion_21/sd21.ipynb:395]:
```suggestion
    "Scale and decode the image latents with VAE."
```

**Comment by kahmed10** [comment]
@attila-dusnoki-htec this looks good. Some minor suggestions above. Would it be possible to add some sort of timing information as well? Such as time it took to run the whole pipeline.

**Comment by causten** [comment]
Can you reduce the png in size?  Perhaps convert to a jpg?   It is currently 420 KB.   Since it will be part of the base repo, the smaller the better

**Comment by attila-dusnoki-htec** [comment]
> Can you reduce the png in size? Perhaps convert to a jpg? It is currently 420 KB. Since it will be part of the base repo, the smaller the better

Sorry about that.
I changed it to jpeg, it is 29K. But i can remove it as well if it is still too large.

**Comment by attila-dusnoki-htec** [comment]
> @attila-dusnoki-htec this looks good. Some minor suggestions above. Would it be possible to add some sort of timing information as well? Such as time it took to run the whole pipeline.

Thanks. I fixed the grammar and added time measurement for certain parts of the code.
The output currently looks like this:  [example_run.log](https://github.com/ROCmSoftwarePlatform/AMDMIGraphX/files/13472287/example_run.log)

**Comment by attila-dusnoki-htec** [comment]
I ended up adding a `gradio` version as well:

gradio_sd_mgx

**Review by kahmed10 — APPROVED** [review]

**Review by causten — APPROVED** [review]
